<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Bioc2022Anno • Bioc2022Anno</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Bioc2022Anno">
<meta property="og:description" content="This is a package for the annotation workshop, presented
	     at BioC 2022.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-93043521-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-93043521-1');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">Bioc2022Anno</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/AnnotationWorkshop.html">Bioconductor 2022 Annotation Workshop</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/jmacdon/Bioc2022Anno" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">

<div class="section level1">
<div class="page-header"><h1 id="introduction-to-bioconductor-annotation-resources">Introduction to Bioconductor annotation resources<a class="anchor" aria-label="anchor" href="#introduction-to-bioconductor-annotation-resources"></a>
</h1></div>
</div>
<div class="section level1">
<h1 id="instructor">Instructor<a class="anchor" aria-label="anchor" href="#instructor"></a>
</h1>
<ul>
<li>James W. MacDonald (<a href="mailto:jmacdon@uw.edu" class="email">jmacdon@uw.edu</a>)</li>
</ul>
</div>
<div class="section level1">
<h1 id="workshop-description">Workshop Description<a class="anchor" aria-label="anchor" href="#workshop-description"></a>
</h1>
<p>There are various annotation packages provided by the Bioconductor project that can be used to incorporate additional information to results from high-throughput experiments. This can be as simple as mapping Ensembl IDs to corresponding HUGO gene symbols, to much more complex queries involving multiple data sources. In this workshop we will cover the various classes of annotation packages, what they contain, and how to use them efficiently.</p>
<div class="section level2">
<h2 id="prerequisites">Prerequisites<a class="anchor" aria-label="anchor" href="#prerequisites"></a>
</h2>
<ul>
<li>Basic knowledge of R syntax</li>
<li>Basic understanding of the various annotation sources (NCBI, EBI/EMBL)</li>
</ul>
<p>Useful background reading</p>
<ul>
<li>The <a href="https://www.bioconductor.org/packages/release/bioc/vignettes/AnnotationDbi/inst/doc/IntroToAnnotationPackages.pdf" class="external-link">AnnotationDbi</a> vignette.</li>
<li>The <a href="https://www.bioconductor.org/packages/release/bioc/vignettes/biomaRt/inst/doc/biomaRt.html" class="external-link">biomaRt</a> vignette.</li>
<li>The <a href="https://www.bioconductor.org/packages/release/bioc/vignettes/GenomicFeatures/inst/doc/GenomicFeatures.pdf" class="external-link">GenomicFeatures</a> vignette.</li>
</ul>
</div>
<div class="section level2">
<h2 id="preparation">Preparation<a class="anchor" aria-label="anchor" href="#preparation"></a>
</h2>
<p>To be able to follow along with this workshop, we have created a Docker installation that includes the devel version of R, and all required Bioconductor packages. In order to access this, you need to first install <a href="https://docs.docker.com/engine/install/" class="external-link">Docker</a> on your own computer. Once you have done that, you can then load the Docker container for this workshop by starting Docker.</p>
<p>How you do this is dependent on your operating system.</p>
<div class="section level3">
<h3 id="windows">Windows<a class="anchor" aria-label="anchor" href="#windows"></a>
</h3>
<p>Hit the ‘Windows’ key (lower left on your keyboard, between Ctrl and Alt), then type Docker. If Docker is installed you should see Docker App highlighted - click Enter to start the App. It can take some time to get started. You can see it’s doing something by clicking on the little caret (^) in the lower right of your screen - there should be a little animated Docker icon which indicates it’s starting. Once it is started, open a CMD prompt by hitting the Windows key again and typing cmd, then Enter. In the CMD prompt type</p>
<pre><code>docker run -e PASSWORD=&lt;choose_a_password_for_rstudio&gt; -p 8787:8787 jmacdon/bioc2022anno</code></pre>
<p>You can choose any password for rstudio - that’s what you will use to log in. It will take some time for the Docker to be downloaded and started, so you might consider doing this ahead of time.</p>
</div>
<div class="section level3">
<h3 id="linux">Linux<a class="anchor" aria-label="anchor" href="#linux"></a>
</h3>
<p>For Linux, it depends on how you installed. If you used a package installer then presumably Docker will be set to start automatically. Otherwise you need to start the Docker daemon by hand (or set it to start automatically). There are too many variables to give much detailed information here; for those on Linux, the assumption is that you probably know what you are doing and can figure it out from the Docker install page.</p>
<p>To start the daemon, if neccessary, do</p>
<pre><code>sudo dockerd &amp;
## followed by 

sudo docker run hello-world
</code></pre>
<p>If Docker is installed correctly it should print something informative. To get the Docker container, it’s the same as for Windows.</p>
<pre><code>docker run -e PASSWORD=&lt;choose_a_password_for_rstudio&gt; -p 8787:8787 jmacdon/bioc2022anno</code></pre>
</div>
<div class="section level3">
<h3 id="macos">MacOS<a class="anchor" aria-label="anchor" href="#macos"></a>
</h3>
<p>There is an installer for MacOS. As with Windows, follow the instructions - it’s just a regular drag’n’drop install. Once it’s installed and started, open a terminal prompt and as above type.</p>
<pre><code>docker run -e PASSWORD=&lt;choose_a_password_for_rstudio&gt; -p 8787:8787 jmacdon/bioc2022anno</code></pre>
<p>For all operating systems, once the Docker container is initialized, you can access it by opening a browser and typing</p>
<pre><code>http://localhost:8787</code></pre>
<p>Which should present you with an RStudio login. Use rstudio as the username and the password you used to start the Docker.</p>
</div>
</div>
<div class="section level2">
<h2 id="workshop-participation">Workshop Participation<a class="anchor" aria-label="anchor" href="#workshop-participation"></a>
</h2>
<p>After each type of annotation package is introduced, students will be given the opportunity to practice making their own queries.</p>
</div>
<div class="section level2">
<h2 id="r--bioconductor-packages-used">
<em>R</em> / <em>Bioconductor</em> packages used<a class="anchor" aria-label="anchor" href="#r--bioconductor-packages-used"></a>
</h2>
<ul>
<li>AnnotationDbi</li>
<li>AnnotationHub</li>
<li>BSgenome</li>
<li>biomaRt</li>
<li>ensembldb</li>
<li>org.Hs.eg.db</li>
<li>TxDb.Hsapiens.UCSC.hg19.knownGene</li>
<li>EnsDb.Hsapiens.v79</li>
<li>EnsDb.Mmusculus.v79</li>
<li>Homo.sapiens</li>
<li>BSgenome.Hsapiens.UCSC.hg19</li>
<li>hugene20sttranscriptcluster.db</li>
</ul>
</div>
</div>
<div class="section level1">
<h1 id="workshop-goals-and-objectives">Workshop goals and objectives<a class="anchor" aria-label="anchor" href="#workshop-goals-and-objectives"></a>
</h1>
<p>Annotating data is a complex task. For any high-throughput experiment the analyst usually starts with a set of identifiers for each thing that was measured, and in order to make the results useful to collaborators these identifiers need to be mapped to other identifiers that are either more familiar to collaborators, or that can be used for further analyses. As an example, RNA-Seq data may only have Entrez Gene IDs for each gene measured, and as part of the output you may want to include the gene symbols, which are more likely to be familiar to a Biologist.</p>
<div class="section level2">
<h2 id="learning-goals">Learning goals<a class="anchor" aria-label="anchor" href="#learning-goals"></a>
</h2>
<ul>
<li>Understand what sort of annotation data are available</li>
<li>Understand the difference between annotation sources (NCBI and EBI/EMBL)</li>
<li>Gain familiarity with the various ways to query annotation packages</li>
<li>Get some practice making queries</li>
</ul>
</div>
<div class="section level2">
<h2 id="learning-objectives">Learning objectives<a class="anchor" aria-label="anchor" href="#learning-objectives"></a>
</h2>
<ul>
<li>Be able to use select and mapIds to map between identifiers</li>
<li>Be able to extract data from TxDb and EnsDb packages</li>
<li>Be able to make queries using biomaRt</li>
<li>Extract and utilize various data from AnnotationHub</li>
</ul>
</div>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/Artistic-2.0" class="external-link">Artistic-2.0</a></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing Bioc2022Anno</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>James MacDonald <br><small class="roles"> Author, maintainer </small>  </li>
<li>Lori Shepherd <br><small class="roles"> Author </small>  </li>
</ul>
</div>



  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by James MacDonald, Lori Shepherd.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
